src/dropout.jl

```diff
@@ -0,0 +1,156 @@
+
+julia> mean(dropout(ones(10^4, 5), 0.3, dims=1), dims=1)
+1×5 Matrix{Float64}:
+ 1.00571  1.00571  1.00571  1.00571  1.00571
+```
+"""
+dropout(A::AbstractArray, p::Real; dims = :) = dropout(_rng_from_array(A), A, p; dims)
+
+function dropout(rng::AbstractRNG, A::AbstractArray, p::Real; dims = :)
+    T = float(eltype(A))
+    0 <= p <= 1 || throw(ArgumentError("dropout expects a probability 0 <= p <= 1"))
+    if p > 0
+        dst = similar(A, T)
+        pT = convert(real(T), p)
+        _dropout!(rng, dst, A, pT, dims)
+    else
+        # Not so sure we want fast paths... this tries but doesn't guarantee type-stability,
+        # and the rrule does not have such a fast paths.
+        convert(AbstractArray{T}, A)
+    end
+end
+
+"""
+    dropout!(B, A, p; dims=:)
+
+This does exactly `B .= dropout(A, p; dims)`,
+or rather, it's the implementation of out-of-place [`dropout`](@ref).
+"""
+function dropout!(dst::AbstractArray, src::AbstractArray, p::Real; dims=:)
+    size(dst) == size(src) || throw(DimensionMismatch("dropout! expects output array the same size as input"))
+    0 <= p <= 1 || throw(ArgumentError("dropout expects a probability 0 <= p <= 1"))
+    if p > 0
+        rng = _rng_from_array(A)
+        pT = convert(real(eltype(dst)), p)
+        _dropout!(rng, dst, src, pT, dims)
+    else
+        # This fast path isn't free, but no concerns about types changing:
+        copyto!(dst, src)
+    end
+end
+
+# This is the easy case in that we can safely use the output array for random numbers.
+function _dropout!(rng::AbstractRNG, dst::AbstractArray, src::AbstractArray, p::Real, dims::Colon)
+    T = real(eltype(dst))
+    val = convert(T, 1/(1-p))
+    rand!(rng, dst)
+    ## This is what we want, but it hits a SIMD bug, solved by _fast_broadcast!
+    # dst .= (dst.>p) .* val .* src
+    _fast_broadcast!(dst, src) do q, x
+        ((real(q)>p) * val) * x
+    end
+    dst
+end
+
+# For other dims, we we do need to allocate something.
+function _dropout!(rng::AbstractRNG, dst::AbstractArray, src::AbstractArray, p::Real, dims)
+    T = real(eltype(dst))
+    tmp = similar(dst, T, ntuple(d -> d in dims ? size(src,d) : 1, ndims(src)))
+    rand!(rng, tmp)
+    val = convert(T, 1/(1-p))
+    ## One-pass strategy -- faster on GPU
+    dst .= ((tmp.>p) .* val) .* src
+    ## Two-pass strategy -- slightly faster on some CPUs?
+    # _fast_broadcast!(tmp) do q
+    #     (q>p) * val
+    # end
+    # dst .= tmp .* src
+end
+
+# The gradient needs to keep the random choices made, thus store at least a BitArray,
+# but the following way turns out to be faster & simpler:
+function ChainRulesCore.rrule(::typeof(dropout), rng::AbstractRNG, A::AbstractArray, p::Real; dims = :)
+    T = float(real(eltype(A)))
+    val = convert(T, 1/(1-p))
+    keep = if dims isa Colon
+        similar(A, T) 
+    else
+        similar(A, T, ntuple(d -> d in dims ? size(A,d) : 1, ndims(A)))
+    end
+    rand!(rng, keep)
+    Y = @. ((keep>p) * val) * A
```

Comment:
Is doing `(keep>p) * val)` twice faster than doing https://github.com/FluxML/NNlib.jl/pull/454/files#diff-e84a46a10c6ff3b32f9a7f735fafee1838628a5261e84f981ec8abcf845547f2R80-R82 once? If so, that's unfortunate.

Reply:
I think it was very close to a wash. This way lets you have just one `back` function, instead of two cases.

Reply:
I don't quite get the point about two `back` functions. Is it not sufficient to just `s/((keep>p) * val)/keep/g`? Zygote might like having to generate fewer `getproperty` calls for captured variables too.

Reply:
On second thought, doing this would mean `dropout` is no longer doubly-differentiable. I retract my previous suggestion.